### PR TITLE
Grand Exchange Plugin

### PR DESCRIPTION
**Pull request by rbbi**
First implementation of Grand Exchange Plugin, some way to go on this - obviously!

TODO:
-  Fix progress bar size and location
-  Add a default icon for empty slots (not dwarf remains [id=0])!
-  Whatever else

image

image

### CODE REVIEW

**Comment by UniquePassive** [comment]
First of all, good work, but please read about our [Coding Format](https://github.com/runelite/runelite/blob/master/CONTRIBUTING.md#-coding-format) (tabs, not spaces).

Please remove the 
```java
/**
 * @author Robbie, created on 29/10/2017 10:34 AM
 */
```
and instead put the license and copyright header at the top of new files.

Lastly, you missed [RSGrandExchangeOffer.java](https://github.com/runelite/runelite/pull/193/files#diff-30651bfd479780faf7e3738040d112d2) when fixing the format in your 2nd commit 😄 (braces on new lines).

That would probably be all to have your code looking good for when you've checked off your to-do.

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java [hunk 1]
```diff
@@ -0,0 +1,625 @@
...
+      populate();
+      component = buildPanel();
+      add(component);
+      handleLogic();
+    }
+  }
+
+  private void populate()
+  {
+    this.offerImages = new ImageIcon[8];
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java]: Consider using an array of objects for these instead of 6 arrays

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+      statusIndicators[i] = new JLabel(); //holds quantity bought/sold
+      progressBars[i] = new JProgressBar();
+    }
+
+  }
+
+  private void handleLogic()
+  {
+    ScheduledExecutorService ses = Executors.newSingleThreadScheduledExecutor();
+    ses.scheduleWithFixedDelay(() -> updateUI.accept(client.getGrandExchangeOffers()), 0, 1000, TimeUnit.MILLISECONDS);
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java]: There is an ```@Scheduled``` annotation to schedule periodic tasks. Although it may be better to hook to the offers field being modified in the client than to poll it.

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+            progressBars[idx].setBackground(decodeColour(progressBars[idx]));
+            progressBars[idx].setForeground(Color.white);
+            progressBars[idx].setString(offer.getQuantitySold() + " / " + offer.getTotalQuantity());
+
+          });
+
+  private String getIsBuySell(GrandExchangeOffer offer)
+  {
+    switch (offer.getProgress()) {
+      case 2:
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java]: These constants would be good to place into GrandExchangeOffer in runelite-api, maybe with some javadoc.

File: runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+
+    JPanel panel_1 = new JPanel();
+    panel_1.setBackground(new Color(90, 82, 69));
+    panel_1.setBorder(new LineBorder(new Color(0, 0, 0)));
+
+    JPanel panel_2 = new JPanel();
+    panel_2.setBorder(new LineBorder(new Color(0, 0, 0)));
+    panel_2.setBackground(new Color(90, 82, 69));
+
+    imageLabels[1] = new JLabel("");
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/grandexchange/GrandExchangePanel.java]: This to line 240 looks duplicated 8 times. Consider moving the shared logic to a method.

File: runescape-api/src/main/java/net/runelite/rs/api/RSClient.java [hunk 5]
```diff
@@ -27,272 +27,272 @@
 import net.runelite.api.Client;
 import net.runelite.mapping.Import;
 
-public interface RSClient extends RSGameEngine, Client
-{
-	@Import("cameraX")
-	@Override
-	int getCameraX();
+public interface RSClient extends RSGameEngine, Client {
+  @Import("cameraX")
```
Comment (Adam-) [inline runescape-api/src/main/java/net/runelite/rs/api/RSClient.java]: Please do not reformat large amounts of unrelated code.

**Comment by rbbi** [comment]
Some progress on this plugin...
- JFrame-related component arrays are stored in their own respective object (GrandExchangeOfferSlot.java)
- Added Apache commons-lang dependency for ImmutablePair
- Rewrote core a bit

Still need to add various labels, fix progress bars, etc.

**Comment by Adam-** [comment]
Thanks for your work on this so far. I don't like the use of Pairs (what is getLeft() and getRight()?) - and certainly don't want to pull in a whole new dependency just for it. Maybe place the GrandExchangeOffer in the GrandExchangeOfferSlot or vise versa. Or, if you must keep the pair-like structure make an object which holds both that is more descriptive.

**Comment by Adam-** [comment]
@rbbi I have cleaned up the plugin to get the design in line with the rest of the plugins. Are you still working on the features outlined in your first comment?

**Comment by rbbi** [comment]
@Adam- thanks for your commits, yes - I've been working my way through the list, also adding new stuff to it that could also prove useful.

**Comment by Adam-** [comment]
Done in #698
